cpp/src/arrow/filesystem/azurefs.cc

```diff
@@ -1066,20 +1111,95 @@ class ObjectAppendStream final : public io::OutputStream {
       // flush. This also avoids some unhandled errors when flushing in the destructor.
       return Status::OK();
     }
-    return CommitBlockList(block_blob_client_, block_ids_, commit_block_list_options_);
+
+    auto fut = FlushAsync();
+    RETURN_NOT_OK(fut.status());
+
+    std::unique_lock<std::mutex> lock(upload_state_->mutex);
+    return CommitBlockList(block_blob_client_, upload_state_->block_ids,
```

Comment:
Indeed, this should happen for both methods. Added it.

Reply:
Thanks for adding it. Do the tests cover this?

Reply:
Yes, this is on the code path that is tested by https://github.com/OliLay/arrow/blob/afa111bc50e2f0ec54c7eb2bdcd6982c155b75a1/cpp/src/arrow/filesystem/azurefs_test.cc#L1529

Reply:
Ok, I ask because it seems the tests didn't catch this? and I happened to spot it.

Reply:
Good point. I think because in the linked test we make sure the destructor is called which will also trigger a commit. 
I wrote an more explicit test which only triggers `CloseAsync()` and makes sure data is written correctly: 48bc6bed5398f062e2f6f04b64873da7186f5f0c